### PR TITLE
Allow user change email

### PR DESCRIPTION
**Pull request by ckouder**
This pull request addresses issue https://github.com/ITI/searcch/issues/155, tested with Google login and github login, haven't tested with cilogon but it should work

### CODE REVIEW

File: searcch_backend/api/resources/login.py [hunk 1]
```diff
@@ -171,7 +172,7 @@ def post(self):
         sso_token = args.get('token')
         login_session = lookup_token(sso_token)
         if not login_session:
-            (user_email, user_name) = self._validate_token(strategy, sso_token)
+            (user_email, user_name, idp_uid) = self._validate_token(strategy, sso_token)
```
Comment (carboxylman) [inline searcch_backend/api/resources/login.py:175]: Let's make sure that the third element of the tuple returned from _validate_token() is stringified, or None.

File: searcch_backend/api/resources/login.py [hunk 2]
```diff
@@ -182,6 +183,24 @@ def post(self):
               filter(Person.email == user_email).\
               first()
 
+            if user:
+                # check if User entity has a UserIdPCredential with this strategy
+                idp_credential = db.session.query(UserIdPCredential).\
+                    filter(UserIdPCredential.user_id == user.id).\
+                    first()
+                setattr(idp_credential, strategy + '_id', idp_uid)
```
Comment (carboxylman) [inline searcch_backend/api/resources/login.py:192]: Let's be careful to insert only strings here.

File: searcch_backend/api/resources/login.py [hunk 3]
```diff
@@ -182,6 +183,24 @@ def post(self):
...
+            else:
+                user = db.session.query(User).\
+                    join(UserIdPCredential, UserIdPCredential.user_id == User.id).\
+                    filter(getattr(UserIdPCredential, strategy + '_id') == idp_uid).\
+                    first()
+                # update user email if it has changed
+                if user:
+                    person = user.person
+                    if person.email != user_email:
+                        person.email = user_email
```
Comment (carboxylman) [inline searcch_backend/api/resources/login.py:204]: Can we please add a DEBUG log statement here, for ease of observability with testing?

File: searcch_backend/api/resources/login.py [hunk 4]
```diff
@@ -201,7 +220,7 @@ def post(self):
                 response.headers.add('Access-Control-Allow-Origin', '*')
                 response.status_code = 200
                 return response
-            else:  # create new user
```
Comment (carboxylman) [inline searcch_backend/api/resources/login.py:204]: Let's leave comments in even if badly formatted, unless they are wrong, which I think this one is not?

File: searcch_backend/migrations/versions/f29a393d7f66_add_user_ipd_credential_table.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+# revision identifiers, used by Alembic.
+revision = 'f29a393d7f66'
+down_revision = 'b11c5baa16f9'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    op.create_table('user_idp_credentials',
```
Comment (carboxylman) [inline searcch_backend/migrations/versions/f29a393d7f66_add_user_ipd_credential_table.py:20]: On the call I suggested that we do something more general, but on further reflection this is good for right now.  Sorry.

**Review by carboxylman — CHANGES_REQUESTED** [review]

**Review by carboxylman — APPROVED** [review]

**Comment by carboxylman** [comment]
@ckouder I added fixes for the existing-user, no-idp_credential case: https://github.com/ITI/searcch-backend/commit/a42631de472e538d6f67343bbfaf12baab32ee1b

**Comment by ckouder** [comment]
> @ckouder I added fixes for the existing-user, no-idp_credential case: [a42631d](https://github.com/ITI/searcch-backend/commit/a42631de472e538d6f67343bbfaf12baab32ee1b)

thank you!

**Pull request merged**